src/charm.py

```diff
@@ -44,21 +44,36 @@ def __init__(self, *args):
         self.framework.observe(self.on['grafana-source'].relation_changed,
                                self._on_grafana_changed)
 
-        if self._stored.provider_ready:
+        if self.provider_ready:
             self.prometheus_provider = MonitoringProvider(self,
                                                           'monitoring', 'prometheus', self.version)
             self.framework.observe(self.prometheus_provider.on.targets_changed,
-                                   self._on_config_changed)
+                                   self._on_scrape_targets_changed)
+
+    def _on_pebble_ready(self, event):
+        """Setup workload container configuration.
+        """
+        self._configure()
 
     def _on_config_changed(self, event):
+        """Handle a configuration change.
+        """
+        self._configure()
+
+    def _on_scrape_targets_changed(self, event):
+        """Handle changes in scrape targets.
+        """
+        self._configure()
+
+    def _configure(self):
         """Reconfigure and restart Prometheus.
 
         In response to any configuration change, such as a new consumer
         relation, or a new configuration set by the administrator, the
         Prometheus config file is regenerated, pushed to the workload
         container and Prometheus is restarted.
         """
-        logger.info("Handling config changed")
+        logger.info("Configuring Prometheus")
```

Comment:
Entrance message should probably be debug. Later on (under config hash change) makes sense to have info.

Reply:
Hmm... Isn't it useful for a system administrator to see that prometheus underwent a configure step when he added a relation. He can use the debug messages if he wants more details.

Reply:
So `_configure` is a common exit hook to many events. You may end up seeing this message even if effectively nothing changed. Guarded `info` messages are more precise.

Reply:
It is true this configure message is going to appear multiple times in the lifecycle of the charm. But each one of those are significant events in the life of the charm. As opposed to say a update status.